### PR TITLE
[ADD] Watermar_image: Adding watermarks in images from products

### PR DESCRIPTION
**Pull request by hmpb-odoo**
The client is a eCommerce, wich add a lot of images. This images has intellectual property. So we add a watermark with the name of the company in every image in product. With this the client ensure that nobody can use this images without autorization.
	Add files:
	product_template(inherith from product.template)
	cron.xml(a aux tool to add watermarks to every image who was alredy in the db)

Task Id: 2944853

### Description

*Please include a summary of the change, the project task link and/or the ID.*

Link to task: [#ID](https://www.odoo.com/web#model=project.task&id=)

### All Submissions:

* [ ] My commit respects the [Odoo commit guideline](https://www.odoo.com/documentation/15.0/developer/misc/other/guidelines.html#git)
* [ ] My commit message respects the [commit template](https://github.com/odoo-ps/psbe-process/wiki/Commits-message-guidelines#template)
* [ ] I have used pre-commit
* [ ] The PR contains **only** my modification and **no other external** commit

### Sh/Runbot:

* [ ] The commits pass test and the branch is green
* [ ] Unit tests have been implemented / standard ones rewritten
* [ ] The Staging is ISO-Prod and will contain only this dev

### Upgrade:

* [ ] The data affected (*if any*) by the changes has been migrated

### CODE REVIEW

File: watermark_image/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    'name': 'Watermark Image',
+    'sumary': 'Add watermark to image in product',
+    'description': """
+        Add watermark to image in product_template and product_image
+        with this the user can ensure that nobody can use the image of the product
+    """,
+    'author': 'Odoo PS',
+    'category': 'Sales',
+    'version': '15.0.1.0.0',
+    'depends': ['sale', 'product', 'website_sale'],
```
Comment (tdetry) [inline watermark_image/__manifest__.py:11]: better on separated lines + you only need website_sale as it include sale and product

File: watermark_image/models/product_template.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        return super().create(vals)
+    
+    def write(self, vals):
+        if vals.get('image_1920', False):
+            vals['image_1920'] = self.add_watermark(vals['image_1920'])
+            
+        if vals.get('product_template_image_ids', False):
+            for product_template_image_id in vals['product_template_image_ids']:
+                if type(product_template_image_id[2]) == dict:
```
Comment (tdetry) [inline watermark_image/models/product_template.py:28]: this looks error-prone / hackish

File: watermark_image/models/product_template.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        return img_str
+
+    def add_watermark_image(self):
+        products = self.env['product.template'].search([])
+        for product in products.filtered(lambda x: x.image_1920):
+                product.image_1920 = self.add_watermark(product.image_1920)
+        
+        products_images = self.env['product.image'].search([])
+        for product_image in products_images.filtered(lambda x: x.image_1920):
+                product_image.image_1920 = self.add_watermark(product_image.image_1920)
```
Comment (tdetry) [inline watermark_image/models/product_template.py:59]: aren't you adding the watermark multiple time with this ? Maybe you need a boolean to mark image as watermarked ?

File: watermark_image/data/cron.xml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    <record id="ir_cron_auto_add_watermark_to_existing_product" model="ir.cron">
+        <field name="name">Product template automated add watermark to existing products</field>
+        <field name="model_id" ref="product.model_product_template"/>
+        <field name="state">code</field>
+        <field name="code">model.add_watermark_image()</field>
+        <field name="interval_number">1</field>
+        <field name="interval_type">minutes</field>
+        <field name="numbercall">1</field>
+    </record>
+</odoo>
```
Comment (tdetry) [inline watermark_image/data/cron.xml:12]: Good, later on you will learn that for data existing before your development you can use migration script, the cron is a bit hackish

File: watermark_image/models/product_template.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+from PIL import ImageDraw, ImageFont, Image
+from io import BytesIO
+import base64
+from odoo import models, api
```
Comment (tdetry) [inline watermark_image/models/product_template.py:4]: separate odoo and standard imports

File: watermark_image/models/product_template.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    @api.model
+    def create(self, vals):
+        if vals.get('image_1920', False):
+            vals['image_1920'] = self.add_watermark(vals['image_1920'])
+    
+        if vals.get('product_template_image_ids', False):
+            for product_template_image_id in vals['product_template_image_ids']:
+                if product_template_image_id[2].get('image_1920', False):
+                    product_template_image_id[2]['image_1920'] = self.add_watermark(product_template_image_id[2]['image_1920'])
```
Comment (tdetry) [inline watermark_image/models/product_template.py:18]: there is duplicated code in create and write, you should create a method for it

File: watermark_image/models/product_template.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        text = "OCHO"
+        font_size = int(height / 4)
+        font = ImageFont.truetype("Ubuntu-R.ttf", font_size)
+        textwidth, textheight = draw.textsize(text, font)
+        x = (width - textwidth) / 2
+        y = (height - textheight) / 2
+        draw.text((x, y), text, font=font, fill=(10))
+        buffered = BytesIO()
+        watermark_image.save(buffered, format="PNG")
+        img_str = base64.b64encode(buffered.getvalue())
```
Comment (tdetry) [inline watermark_image/models/product_template.py:49]: I see you had fun :D maybe not the most optimized in terms of memory but if it works smoothly.. ok ;-)

**Review by tdetry — CHANGES_REQUESTED** [review]
some improvements needed ;-)

File: watermark_image/models/product_template.py (hunk 3 above)
Comment (hmpb-odoo) [reply watermark_image/models/product_template.py]: done

File: watermark_image/models/product_template.py (hunk 7 above)
Comment (hmpb-odoo) [reply watermark_image/models/product_template.py]: It was great, I wish I had more time to do a fancy design for the watermark :)

**Pull request closed without merging**